lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,394 @@
+
+  return success();
+}
+
+// Lowering connects only has to deal with one special case: connecting two
+// bundles. This situation should only arise when both of the arguments are a
+// bundle that was:
+//   a) originally a block argument
+//   b) originally an instance's port
+//
+// When two such bundles are connected, none of the subfield visits have a
+// chance to lower them, so we must ensure they have the same number of
+// flattened values and flatten out this connect into multiple connects.
+LogicalResult FIRRTLTypesLowering::visitStmt(ConnectOp op) {
+  Value dest = op.dest();
+  Value src = op.src();
+
+  // If we aren't connecting two bundles, there is nothing to do.
+  if (!dest.getType().isa<BundleType>() || !src.getType().isa<BundleType>())
+    return success();
+
+  // Get the lowered values for each side.
+  SmallVector<Value, 8> destValues;
+  getAllBundleLowerings(dest, destValues);
+
+  SmallVector<Value, 8> srcValues;
+  getAllBundleLowerings(src, srcValues);
+
+  // Check that we got out the same number of values from each bundle.
+  assert(destValues.size() == srcValues.size() &&
+         "connected bundles don't match");
+
+  for (auto tuple : llvm::zip_first(destValues, srcValues)) {
+    Value newDest = std::get<0>(tuple);
+    Value newSrc = std::get<1>(tuple);
+    builder->create<ConnectOp>(op.getLoc(), newDest, newSrc);
+  }
+
+  removeOpLater(op);
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Helpers to manage state.
+//===----------------------------------------------------------------------===//
+
+static DictionaryAttr getArgAttrs(StringAttr nameAttr, StringRef suffix,
+                                  OpBuilder *builder) {
+  SmallString<16> newName(nameAttr.getValue());
+  newName += suffix;
+
+  StringAttr newNameAttr = builder->getStringAttr(newName);
+  auto attr = builder->getNamedAttr("firrtl.name", newNameAttr);
+
+  return builder->getDictionaryAttr(attr);
+}
+
+// Creates and returns a new block argument of the specified type to the module.
+// This also maintains the name attribute for the new argument, possibly with a
+// new suffix appended.
+Value FIRRTLTypesLowering::addArg(Type type, unsigned oldArgNumber,
+                                  StringRef nameSuffix) {
+  FModuleOp module = getOperation();
+  Block *body = module.getBodyBlock();
+
+  // Append the new argument.
+  auto newValue = body->addArgument(type);
+
+  // Keep the module's type up-to-date.
+  module.setType(builder->getFunctionType(body->getArgumentTypes(), {}));
+
+  // Copy over the name attribute for the new argument.
+  StringAttr nameAttr = getFIRRTLNameAttr(module.getArgAttrs(oldArgNumber));
+  if (nameAttr) {
+    auto newArgAttrs = getArgAttrs(nameAttr, nameSuffix, builder);
+    module.setArgAttrs(newValue.getArgNumber(), newArgAttrs);
+  }
+
+  // Remember to delete the original block argument.
+  removeArgLater(oldArgNumber);
```

Comment:
I think this is adding an Argument to the argsToRemove set multiple times, one for each element that gets lowered.  I'd move the this to the "lowerArg" method.  I'd also inline the removeArgLater method as the abstraction make the code clearer.